File: eLit/Model/DrinkParts/Ingredient.swift
//
//  Ingredient.swift
//  eLit
//
//  Created by Gianpaolo Di Pietro on 29/12/2018.
//  Copyright © 2018 Alberto Mario Bellini. All rights reserved.
//

import UIKit
import CoreData

@objc(Ingredient)
class Ingredient: DrinkObjectWithImage {
    
    //MARK: Initializers
    convenience init(grade: Double, name: String) {
        self.init()
        self.grade = grade
        self.name = name
    }
    
    convenience init(name: String) {
        self.init(grade: 0, name: name)
    }
    
    convenience init(dict: [String: Any]) {
        self.init()
        self.id = dict["id"] as? String ?? ""
        self.fingerprint = dict["fingerprint"] as? String ?? ""
        self.grade = dict["grade"] as? Double ?? 0.0
        self.name = dict["name"] as? String
        self.ingredientDescription = dict["ingredient_description"] as? String
        self.imageURLString = dict["image"] as? String ?? ""
    }
    
    override func update(with data: [String : Any], savePersistent: Bool) {
        self.grade = data["grade"] as? Double ?? 0.0
        self.name = data["name"] as? String
        self.ingredientDescription = data["ingredient_description"] as? String
        self.imageURLString = data["image"] as? String ?? ""
        self.image = nil
        self.imageData = nil
        super.update(with: data, savePersistent: savePersistent)
    }

}

File: eLitServer/source/database_classes/ingredient.py
from typing import Dict

import mongoengine as me
from database_classes.db_object import DBObject
from database_classes.drink_object import DrinkObject


class Ingredient(DrinkObject):
    name = me.StringField(required=True, unique=True)
    grade = me.FloatField()
    image = me.StringField()
    ingredient_description = me.StringField()

    def __init__(self, name: str, grade: float = 0, image: str = '', ingredient_description: str = '', *args, **values):
        super().__init__(*args, **values)
        self.name = name
        self.grade = grade
        self.ingredient_description = ingredient_description
        self.image = image

    def to_dict(self) -> Dict:
        data = super().to_dict()
        data['name'] = self.name
        data['grade'] = self.grade
        data['ingredient_description'] = self.ingredient_description
        data['image'] = self.image
        return data
